### ISSUE TITLE
DataStore and API - Crashes when having unstable or intermittent Internet connection.

### ISSUE THREAD

**Issue by upachecog** [aws-amplify/amplify-android#1132]
**Short Description:** When using a low-end device with a really poor Internet connection I sometimes have some crashes when using the API and DataStore categories.

**Version:** Datastore and API 1.6.9

**Related code:**

This is the first call in my App, just a simple query to the API:

```kotlin
Amplify.API.query(
    ModelQuery.list(MyModel::class.java, MyModel.CREATED_AT.contains(date)),
    { response -> onResponse(response.data.toList()) },
    onFailure
)
```

Then I call DataStore start:

```kotlin
Amplify.DataStore.start(
    { onSuccess() },
    { Amplify.DataStore.start(
        { onSuccess() },
        { failure -> onFailure(failure) }
    ) }
)
``` 
`onSuccess()` navigates to a new Activity.

After a successful response I call a query to the DataStore:

```kotlin
Amplify.DataStore.query(
    MyModel::class.java,
     Where.matches(MyModel.DATE.eq(Utils.getCurrentTemporalDate())),
     { onResponse() },
     { onFailure() }
)
```

4. Here are some logs I got from Sentry and that I think could be related:

https://sentry.io/share/issue/e5b30629b5654f96806593bde3b23b3b/
https://sentry.io/share/issue/9feaa777a3a146a0938ccf34d88fb4f4/
https://sentry.io/share/issue/82dded9be9e247d887229538cfebcbb5/
https://sentry.io/share/issue/2642b63570924866b2a68c4bfd713d10/
https://sentry.io/share/issue/605db6748b494cbfa786257ff29666a8/

**Comment by richardmcclellan** [comment]
Hey @upachecog, using the `DataStore` requires conflict detection to be _enabled_ on the AppSync API.  Using `API` requires conflict detection to be _disabled_.   For this reason, there's not an easy way to use `DataStore`, and `API` in the same application.  Are you trying to use both `API` and `DataStore` in the same application?

Also, FYI, `Amplify.DataStore.query` will automatically call `Amplify.DataStore.start`, if needed, so calling `Amplify.DataStore.start` is not explicitly needed, unless you want the sync to be started ahead of time, such as before you are ready to display any data.

**Comment by upachecog** [comment]
Hi @richardmcclellan, yes, but just that query is calling the API. Let me update that call and I'll let you know if the issue is still there.

Thank you for your observation.

**Comment by upachecog** [comment]
For this feature, I require the latest information of a Questionnaire and Response models. So I must wait until the DataStore has synced these two models successfully.

I observing the DataStore and starting the DataStore:

```
    private val readyEventSubscription = Amplify.Hub.subscribe(
        HubChannel.DATASTORE,
        { hubEvent -> DataStoreChannelEventName.MODEL_SYNCED.toString() == hubEvent.name },
        { hubEvent ->
            val event = hubEvent.data as ModelSyncedEvent
            when (event.model) {
                QUESTIONNAIRE_MODEL_NAME -> {
                    hasSyncedQuestionnaire = true
                    validateModelSync()
                }
                RESPONSE_QUESTIONNAIRE_MODEL_NAME -> {
                    hasSyncedResponse = true
                    validateModelSync()
                }
            }
        }
    )
  
    init {
       fun startDataStore(
           onResponse: () -> Unit,
           onFailure: () -> Unit
       ) {
           Amplify.DataStore.start({ onResponse() }, { onFailure() })
       }
   }

```

Then I do my validation, if both models have been synced I call my first query:

```
    private fun validateModelSync() {
        if (hasSyncedQuestionnaire && hasSyncedResponse) {
            callRequestTodayQuestionnaire()
            unsubscribe()
        }
    }
```

Then I do the following queries:

```
    private fun callRequestTodayQuestionnaire() {
        val todayDate = Utils.getCurrentFormattedISODateString()
        Log.i(TAG, "Executed query")
        Amplify.DataStore.query(
            Responses::class.java,
            Where.matches(Responses.CREATED_AT.contains(todayDate)),
            { matches ->
                /** If hasNext it means today's questionnaire is answered */
                if (matches.hasNext()) {
                    onResponseTodayQuestionnaire?.let { it(matches.asSequence().toList()) }
                } else {
                    callQuestionnaire()
                }
            },
            { onFailureTodayQuestionnaire?.let { failure -> failure(it) } }
        )
    }

    private fun callQuestionnaire() {
        Amplify.DataStore.query(
            Questionnaire::class.java,
            { matches -> onResponseQuestionnaire?.let { it(matches.asSequence().toList()) } },
            { onFailureQuestionnaire?.let { failure -> failure(it) } }
        )
    }
```

I tried turning off the WiFi at different moments of that process and I'm still facing the same issues. 

Here you can find some logs:

[Amplify-Issue-1132.log](https://github.com/aws-amplify/amplify-android/files/5871936/Amplify-Issue-1132.log)

**Comment by richardmcclellan** [comment]
Hey @upachecog, this is the main error I see in your logs: 
```
javax.net.ssl.SSLException: Read error: ssl=0xa52cc508: I/O error during system call, Software caused connection abort
```

This indicates a lack of network connection (likely caused by turning off WiFi).

Whenever there is a connection error, DataStore is automatically stopped.  `DataStore.query` is a local operation - it simply returns what is in the local database, so any API sync errors are not reported in the `onFailure` callback.  Instead, you can listen for failures by subscribing to the [DataStore `networkStatus` event](https://docs.amplify.aws/lib/datastore/datastore-events/q/platform/android#networkstatus) and looking for a payload with `active: false`.  When you receive this event, I suggest you display a Retry button in your app, which calls `Amplify.DataStore.start` on click, that your end user can press once their device has a connection again.

**Comment by upachecog** [comment]
Hello, @richardmcclellan I implemented a solution as you suggested.

Unfortunately, I am still finding some cases when the application crashes.

**Case 1:** After several times retrying without turning the Internet On.

```
E/AndroidRuntime: FATAL EXCEPTION: RxCachedThreadScheduler-42
    Process: com.invernaderolabs.checkin, PID: 18613
    io.reactivex.rxjava3.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | DataStoreException{message=Error during subscription., cause=ApiException{message=Connection failed., cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}, recoverySuggestion=Evaluate details.}
        at io.reactivex.rxjava3.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:367)
        at io.reactivex.rxjava3.internal.operators.completable.CompletableCreate$Emitter.onError(CompletableCreate.java:78)
        at io.reactivex.rxjava3.internal.operators.completable.CompletableCreate.subscribeActual(CompletableCreate.java:43)
        at io.reactivex.rxjava3.core.Completable.subscribe(Completable.java:2850)
        at io.reactivex.rxjava3.internal.operators.completable.CompletablePeek.subscribeActual(CompletablePeek.java:51)
        at io.reactivex.rxjava3.core.Completable.subscribe(Completable.java:2850)
        at io.reactivex.rxjava3.internal.operators.completable.CompletablePeek.subscribeActual(CompletablePeek.java:51)
        at io.reactivex.rxjava3.core.Completable.subscribe(Completable.java:2850)
        at io.reactivex.rxjava3.internal.operators.completable.CompletablePeek.subscribeActual(CompletablePeek.java:51)
        at io.reactivex.rxjava3.core.Completable.subscribe(Completable.java:2850)
        at io.reactivex.rxjava3.internal.operators.completable.CompletableSubscribeOn$SubscribeOnObserver.run(CompletableSubscribeOn.java:64)
        at io.reactivex.rxjava3.core.Scheduler$DisposeTask.run(Scheduler.java:614)
        at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:65)
        at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:56)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: DataStoreException{message=Error during subscription., cause=ApiException{message=Connection failed., cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}, recoverySuggestion=Evaluate details.}
        at com.amplifyframework.datastore.appsync.AppSyncClient.lambda$subscription$3(AppSyncClient.java:298)
        at com.amplifyframework.datastore.appsync.-$$Lambda$AppSyncClient$797ziDK0io-qXODzROLOA77stS8.accept(Unknown Source:4)
        at com.amplifyframework.api.aws.SubscriptionOperation.lambda$null$1$SubscriptionOperation(SubscriptionOperation.java:92)
        at com.amplifyframework.api.aws.-$$Lambda$SubscriptionOperation$l9lC7Ew6niNmyszMXQSUKS6YhwE.accept(Unknown Source:4)
        at com.amplifyframework.api.aws.SubscriptionEndpoint.requestSubscription(SubscriptionEndpoint.java:127)
        at com.amplifyframework.api.aws.SubscriptionOperation.lambda$start$2$SubscriptionOperation(SubscriptionOperation.java:83)
        at com.amplifyframework.api.aws.-$$Lambda$SubscriptionOperation$dV6kPdbE8Z12Qd0_rE-16cujgUk.run(Unknown Source:2)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:458)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:764) 
     Caused by: ApiException{message=Connection failed., cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}
        at com.amplifyframework.api.aws.SubscriptionEndpoint.requestSubscription(SubscriptionEndpoint.java:128)
        at com.amplifyframework.api.aws.SubscriptionOperation.lambda$start$2$SubscriptionOperation(SubscriptionOperation.java:83) 
        at com.amplifyframework.api.aws.-$$Lambda$SubscriptionOperation$dV6kPdbE8Z12Qd0_rE-16cujgUk.run(Unknown Source:2) 
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:458) 
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:764) 
```

**Case 2:** While the DataStore is starting it crashes when I turn the WiFi off: 
(This case just crashes the app some times.)

```
E/amplify:aws-datastore: Failure encountered while attempting to start API sync.
    DataStoreException{message=Error during subscription., cause=ApiException{message=Connection failed., cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}, recoverySuggestion=Evaluate details.}
        at com.amplifyframework.datastore.appsync.AppSyncClient.lambda$subscription$3(AppSyncClient.java:298)
        at com.amplifyframework.datastore.appsync.-$$Lambda$AppSyncClient$797ziDK0io-qXODzROLOA77stS8.accept(Unknown Source:4)
        at com.amplifyframework.api.aws.SubscriptionOperation.lambda$null$1$SubscriptionOperation(SubscriptionOperation.java:92)
        at com.amplifyframework.api.aws.-$$Lambda$SubscriptionOperation$l9lC7Ew6niNmyszMXQSUKS6YhwE.accept(Unknown Source:4)
        at com.amplifyframework.api.aws.SubscriptionEndpoint.requestSubscription(SubscriptionEndpoint.java:127)
        at com.amplifyframework.api.aws.SubscriptionOperation.lambda$start$2$SubscriptionOperation(SubscriptionOperation.java:83)
        at com.amplifyframework.api.aws.-$$Lambda$SubscriptionOperation$dV6kPdbE8Z12Qd0_rE-16cujgUk.run(Unknown Source:2)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:458)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: ApiException{message=Connection failed., cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}
        at com.amplifyframework.api.aws.SubscriptionEndpoint.requestSubscription(SubscriptionEndpoint.java:128)
        at com.amplifyframework.api.aws.SubscriptionOperation.lambda$start$2$SubscriptionOperation(SubscriptionOperation.java:83) 
        at com.amplifyframework.api.aws.-$$Lambda$SubscriptionOperation$dV6kPdbE8Z12Qd0_rE-16cujgUk.run(Unknown Source:2) 
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:458) 
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:764) 
```

**Comment by richardmcclellan** [comment]
Hi @upachecog,

**Case 1** sounds like https://github.com/aws-amplify/amplify-android/issues/541, which was resolved in 1.6.8.  Are you using 1.6.8 or later?   If so, can you include more of the logcat output leading up to the crash?

**Case 2** sounds like expected behavior.  `DataStore.start()` will attempt to synchronize with the backend, but if there is no network connection, it will automatically be stopped.

**Comment by joekiller** [comment]
Hi I have fixed this with https://github.com/aws-amplify/amplify-android/pull/1755
